### PR TITLE
ci: add feature matrix and --all-features test coverage

### LINKED ISSUE #751: [CI]: Add feature matrix and --all-features test coverage
## Motivation

While #736 ensures `cargo check --workspace --all-targets --all-features` runs in CI, test coverage across feature configurations is still limited.

Feature-gated code and optional dependencies can compile successfully but remain untested under certain configurations. Over time, this may lead to feature drift or regressions in non-default builds.

## Proposal

Extend CI coverage to validate both compilation and test execution across feature combinations:

1. Add:
   cargo test --workspace --all-features

2. Introduce a feature matrix strategy covering:
   - Default features
   - All features
   - Minimal features (`--no-default-features`)

This ensures:
- Feature-gated test paths are exercised
- Optional dependencies remain validated
- Both minimal and maximal configurations remain healthy

Happy to prepare a focused PR if this direction aligns with project goals.

### 📂 Feature Category

Rust Core Library

### 💡 Motivation

Currently, CI validates compilation across all features (via #736), but it does not execute tests across feature combinations.

This creates a potential blind spot where:
- Feature-gated test code may silently break
- Optional dependencies may compile but fail at runtime
- Minimal builds (`--no-default-features`) may drift over time

Improving CI coverage across feature configurations strengthens reliability and long-term maintainability.

### 💭 Proposed Solution

Update the existing GitHub Actions workflow to:

1. Add a dedicated step for:
   cargo test --workspace --all-features

2. Introduce a job matrix similar to:

   - default
   - all-features
   - no-default-features

Each job would run:
   cargo check
   cargo test

Keeping them as separate CI steps makes failures easier to diagnose and avoids conflating feature-related issues.

### 📚 Library Reference

This approach follows common Rust CI best practices used in projects such as:

- tokio
- serde
- clap

Many mature Rust crates validate default, minimal, and full feature sets to prevent feature drift and optional dependency breakage.

### 🔄 Alternatives Considered

An alternative would be to rely solely on `cargo check --all-features`, but this does not execute test code.

Another option is to selectively test only certain feature combinations; however, a small matrix approach provides stronger guarantees while keeping CI manageable.

### 🎯 Use Cases

- Ensuring optional dependencies remain functional.
- Preventing regressions in `--no-default-features` builds.
- Validating feature-gated test modules.
- Increasing confidence for downstream users integrating kornia-rs in custom configurations.

### 📝 Additional Context

_No response_

### 🤝 Contribution Intent

- [x] I plan to submit a PR to implement this feature
- [ ] I'm requesting this feature but not planning to implement it

### PR DESCRIPTION
**Pull request by Ayushi6002**
Closes #751
## 📝 Description

**⚠️ Issue Link Required**: This PR must be linked to an approved and assigned issue. See [Contributing Guide](CONTRIBUTING.md#pull-request) for details.

**Fixes/Relates to:** # (issue number)

**Important**:
- Ensure you are assigned to the linked issue before submitting this PR
- This PR should strictly implement what the linked issue describes
- Do not include changes beyond the scope of the linked issue

---

## 🛠️ Changes Made
- [ ] Item 1
- [ ] Item 2

---

## 🧪 How Was This Tested?
- [ ] **Unit Tests:** (List new/updated tests)
- [ ] **Manual Verification:** (Describe the steps you took)
- [ ] **Performance/Edge Cases:** (How does this handle nulls, large data, etc.?)

---

## 🕵️ AI Usage Disclosure
*Check one of the following:*
- [x] 🟢 **No AI used.**
- [ ] 🟡 **AI-assisted:** I used AI for boilerplate/refactoring but have manually reviewed and tested every line.
- [ ] 🔴 **AI-generated:** (Note: These PRs may be subject to stricter scrutiny or immediate closure if the logic is not explained).

---

## 🚦 Checklist
- [x] I am assigned to the linked issue (required before PR submission)
- [x] The linked issue has been approved by a maintainer
- [x] This PR strictly implements what the linked issue describes (no scope creep)
- [ ] I have performed a **self-review** of my code (no "ghost" variables or hallucinations).
- [x] My code follows the existing style guidelines of this project.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] (Optional) I have attached screenshots/recordings for UI changes.

---

## 💭 Additional Context
Add any other context or screenshots about the pull request here.

### CODE REVIEW

**Comment by sidd-27** [comment]
Using cargo hack here might help

File: .github/workflows/rust_test.yml [hunk 1]
```diff
@@ -53,8 +71,13 @@ jobs:
...
       - name: Run ${{ matrix.command }}
-        run: pixi run ${{ matrix.command }}
-
+        run: |
+          if [ -z "${{ matrix.features }}" ]; then
+            pixi run ${{ matrix.command }}
+          else
+            pixi run ${{ matrix.command }} -- ${{ matrix.features }}
+          fi
+
```
Comment (sidd-27) [inline .github/workflows/rust_test.yml]: cargo hack would be cleaner and more maintainable

**Review by sidd-27 — CHANGES_REQUESTED** [review]

File: .github/workflows/rust_test.yml [hunk 2]
```diff
@@ -25,20 +25,38 @@ jobs:
   test:
     name: ${{ matrix.name }}
     runs-on: ${{ matrix.runner }}
-    timeout-minutes: 10
+    timeout-minutes: 15
```
Comment (sidd-27) [inline .github/workflows/rust_test.yml]: why has this been changed?

File: .github/workflows/rust_test.yml [hunk 3]
```diff
@@ -52,10 +70,30 @@ jobs:
...
+            fi
+          else
+            if [ "${{ matrix.mode }}" = "default" ]; then
+              pixi run cargo test --workspace
+            else
+              pixi run cargo hack test --workspace \
+                $([ "${{ matrix.mode }}" = "all-features" ] && echo "--all-features") \
+                $([ "${{ matrix.mode }}" = "no-default-features" ] && echo "--no-default-features")
+            fi
+          fi
```
Comment (sidd-27) [inline .github/workflows/rust_test.yml]: this feels very hacky,

```
jobs:
  test:
    runs-on: ubuntu-latest
    strategy:
      matrix:
        # Define your logical modes
        mode: [default, all-features, no-default-features]
        release: [true, false]
        include:
          # Map each mode to its specific CLI flags
          - mode: default
            flags: ""
          - mode: all-features
            flags: "--all-features"
          - mode: no-default-features
            flags: "--no-default-features"
          # Map release boolean to the CLI flag
          - release: true
            rel_flag: "--release"
          - release: false
            rel_flag: ""
```
map strategies like this and just run them all using 

```
- name: Run tests (${{ matrix.mode }}, release=${{ matrix.release }})
        run: |
          pixi run cargo hack test --workspace ${{ matrix.flags }} ${{ matrix.rel_flag }}
```

File: .github/workflows/rust_test.yml [hunk 4]
```diff
@@ -52,10 +70,30 @@ jobs:
           pixi-version: v0.40.0
           cache: true
 
-      - name: Run ${{ matrix.command }}
-        run: pixi run ${{ matrix.command }}
+      - name: Install cargo-hack
+        run: pixi run cargo install cargo-hack
```
Comment (sidd-27) [inline .github/workflows/rust_test.yml]: this would be really slow, just add cargo hack to the pixi toml

**Review by sidd-27 — CHANGES_REQUESTED** [review]

**Comment by sidd-27** [comment]
are you still interested in working on this?

**Comment by namanguptagit** [comment]
Can i work on it if its avialable to work on @sidd-27 ?

**Comment by Ayushi6002** [comment]
> are you still interested in working on this?

Yes, I'm still working on it. I'll update the matrix structure as suggested and push changes soon. Thanks for the feedback!

**Comment by Ayushi6002** [comment]
Updated the matrix strategy to use flag mapping as suggested and removed conditional logic. Also integrated cargo-hack via pixi instead of installing in workflow. Please let me know if further refinements are needed.

**Review by sidd-27 — APPROVED** [review]

**Comment by sidd-27** [comment]
@Ayushi6002  fix ci tests

**Comment by Ayushi6002** [comment]
> @Ayushi6002 fix ci tests

@sidd-27 I've pushed the trailing whitespace fix for rust_test.yml. I also noticed all CI failures are due to pixi install --locked failing — this seems to be a pre-existing pixi.lock sync issue unrelated to my changes. Could you please approve the workflow run and help with the lock file if needed?

**Comment by sidd-27** [comment]
run pixi update locally and push i think that should solve it

**Comment by Ayushi6002** [comment]
> run pixi update locally and push i think that should solve it

@sidd-27 I'm on Windows and WSL requires BIOS virtualization which I haven't enabled before. Could you help me with this or suggest an alternative? 🙏

**Comment by sidd-27** [comment]
hmm there seems to be some issues with the cuda stuff

**Comment by Ayushi6002** [comment]
Hi @sidd-27 👋
Just checking in to keep this PR active. I'm still interested in completing this — could you let me know if there's anything specific I should fix on my end to move it forward? Happy to address any remaining issues.
Thanks!

**Comment by sidd-27** [comment]
> specific I should fix on my end to move it forward

if you can fix the ci tests then sure
